### PR TITLE
optimize `TLSConfig` to `addSafeTLSToYaml`

### PR DESCRIPTION
**Pull request by dongjiang1989**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

```Could we do the same thing that we do with ProxyConfig etc,. and just handle the nil value of TLSConfig within the addSafeTLSToYaml ?```
Originally posted by @mviswanathsai in https://github.com/prometheus-operator/prometheus-operator/pull/6711#discussion_r1658394339

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [x] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
optimize `TLSConfig` to `addSafeTLSToYaml`
```

### CODE REVIEW

File: pkg/assets/interface.go [hunk 1]
```diff
@@ -26,4 +26,5 @@ type StoreGetter interface {
 	GetSecretOrConfigMapKey(key monitoringv1.SecretOrConfigMap) (string, error)
 	GetConfigMapKey(key v1.ConfigMapKeySelector) (string, error)
 	GetSecretKey(key v1.SecretKeySelector) ([]byte, error)
+	GetNamespace() string
```
Comment (simonpasquier) [inline pkg/assets/interface.go:29]: (suggestion) we could make TLSAsset() part of the StoreGetter interface

TLSAsset(key interface{}) string
```suggestion
	TLSAsset(key interface{}) string
```

Comment (dongjiang1989) [reply pkg/assets/interface.go:29]: Hmmm.... Maybe make`GetNamespace() string` part of the StoreGetter interface is good.
When `s := store.ForNamespace(ns)`, StoreGetter can not get namespace.
```golang
// ForNamespace returns a StoreGetter scoped to the given namespace.
// It reads data only from the cache which needs to be populated beforehand.
// The namespace argument can't be empty.
func (s *StoreBuilder) ForNamespace(namespace string) StoreGetter {
	if namespace == "" {
		panic("namespace can't be empty")
	}
	return &cacheOnlyStore{
		ns: namespace,
		c:  s.objStore,
	}
}

type cacheOnlyStore struct {
	ns string
	c  cache.Store
}
```

Comment (simonpasquier) [reply pkg/assets/interface.go:29]: My idea was to adapt the existing TLSAsset() function like this:

```golang
func (cos *cacheOnlyStore) TLSAsset(sel interface{}) string {
    var k tlsAssetKey

    switch v := sel.(type) {
    case monitoringv1.SecretOrConfigMap:
        k = tlsAssetKeyFromSelector(cos.ns, v)
    case *v1.SecretKeySelector:
        k = tlsAssetKeyFromSecretSelector(cos.ns, v)
    default:
        return ""
    }

    return k.toString()
}

```

Comment (dongjiang1989) [reply pkg/assets/interface.go]: I got it. Thanks.

File: pkg/prometheus/store.go [hunk 2]
```diff
@@ -56,6 +56,10 @@ func AddRemoteWritesToStore(ctx context.Context, store *assets.StoreBuilder, nam
 		if err := store.AddProxyConfig(ctx, namespace, remote.ProxyConfig); err != nil {
 			return fmt.Errorf("remote write %d: %w", i, err)
 		}
+
+		if err := store.AddTLSConfig(ctx, namespace, remote.TLSConfig); err != nil {
```
Comment (simonpasquier) [inline pkg/prometheus/store.go:60]: it's already called above (at L40)?

File: pkg/prometheus/store.go [hunk 3]
```diff
@@ -99,6 +107,10 @@ func AddAPIServerConfigToStore(ctx context.Context, store *assets.StoreBuilder,
 		return fmt.Errorf("apiserver config: %w", err)
 	}
 
+	if err := store.AddTLSConfig(ctx, namespace, config.TLSConfig); err != nil {
```
Comment (simonpasquier) [inline pkg/prometheus/store.go:110]: Good catch here!

File: pkg/prometheus/store.go [hunk 4]
```diff
@@ -82,6 +86,10 @@ func AddRemoteReadsToStore(ctx context.Context, store *assets.StoreBuilder, name
 		if err := store.AddProxyConfig(ctx, namespace, remote.ProxyConfig); err != nil {
 			return fmt.Errorf("remote read %d: %w", i, err)
 		}
+
+		if err := store.AddTLSConfig(ctx, namespace, remote.TLSConfig); err != nil {
```
Comment (simonpasquier) [inline pkg/prometheus/store.go:90]: same here

File: pkg/prometheus/promcfg.go [hunk 5]
```diff
@@ -1798,7 +1807,7 @@ func (cg *ConfigGenerator) generateK8SSDConfig(
 
 		// TODO: If we want to support secret refs for k8s service discovery tls
 		// config as well, make sure to path the right namespace here.
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1809]: Hmm not sure what the TODO comment is about. I guess that we can remove it.

File: pkg/prometheus/promcfg.go [hunk 6]
```diff
@@ -1846,7 +1855,7 @@ func (cg *ConfigGenerator) generateAlertmanagerConfig(alerting *monitoringv1.Ale
 
 		// TODO: If we want to support secret refs for alertmanager config tls
 		// config as well, make sure to path the right namespace here.
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1857]: same here

**Review by simonpasquier — COMMENTED** [review]
IIUC we have a similar issue with addAlertmanagerEndpointsToStore() not calling `store.AddTLSConfig`()?

File: pkg/prometheus/store.go (hunk 2 above)
Comment (dongjiang1989) [reply pkg/prometheus/store.go:60]: Thanks. My bad. 
Removed it done.

File: pkg/prometheus/store.go (hunk 4 above)
Comment (dongjiang1989) [reply pkg/prometheus/store.go:90]: Thanks. My bad.
Removed it done.

File: pkg/prometheus/promcfg.go (hunk 6 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:1857]: Got it. Remove it done.

File: pkg/prometheus/promcfg.go (hunk 5 above)
Comment (dongjiang1989) [reply pkg/prometheus/promcfg.go:1809]: Got it. Remove it done.

File: pkg/prometheus/store.go [hunk 7]
```diff
@@ -110,3 +114,25 @@ func AddScrapeClassesToStore(ctx context.Context, store *assets.StoreBuilder, na
 	}
 	return nil
 }
+
+func AddAlertmanagerEndpointsToStore(ctx context.Context, store *assets.StoreBuilder, namespace string, ams []monv1.AlertmanagerEndpoints) error {
```
Comment (simonpasquier) [inline pkg/prometheus/store.go:118]: no need to move it to pkg/prometheus because it's only used by the server controller.

**Comment by dongjiang1989** [comment]
> IIUC we have a similar issue with addAlertmanagerEndpointsToStore() not calling `store.AddTLSConfig`()?

Thanks. Got it.
PTAL re-check. @simonpasquier

File: pkg/prometheus/store.go (hunk 7 above)
Comment (dongjiang1989) [reply pkg/prometheus/store.go:118]: Got it. Thanks.

Comment (dongjiang1989) [reply pkg/prometheus/store.go]: @simonpasquier  PTAL, thanks.

**Comment by dongjiang1989** [comment]
Thanks. Please re-check it. @simonpasquier

**Comment by ArthurSens** [comment]
With Simon on PTO, do you mind taking a look @mviswanathsai? I'm not really used to this part of the code, it might take a while for me to catch up

**Comment by dongjiang1989** [comment]
@mviswanathsai PTAL, thanks

**Comment by mviswanathsai** [comment]
I'm sorry I completely forgot. I will have a look ASAP. thanks for reminding me.

File: pkg/prometheus/server/operator.go [hunk 8]
```diff
@@ -1318,6 +1318,10 @@ func addAlertmanagerEndpointsToStore(ctx context.Context, store *assets.StoreBui
 		if err := store.AddSigV4(ctx, namespace, am.Sigv4); err != nil {
 			return fmt.Errorf("alertmanager %d: %w", i, err)
 		}
+
+		if err := store.AddTLSConfig(ctx, namespace, am.TLSConfig); err != nil {
```
Comment (mviswanathsai) [inline pkg/prometheus/server/operator.go:1322]: Curious... didn't we need to add the TLS config to cache even before this PR, Did we miss that completely?

**Comment by mviswanathsai** [comment]
LGTM overall.

**Review by mviswanathsai — APPROVED** [review]
LGTM, I would still ask for @slashpai 's opinion though.

**Comment by mviswanathsai** [comment]
Could you also add nil checks in `addSafeTLStoYaml`? and skip checking for nil TLS outside of it? You can also do it in a follow-up PR if you like. WDYT?

File: pkg/prometheus/server/operator.go (hunk 8 above)
Comment (dongjiang1989) [reply pkg/prometheus/server/operator.go:1322]: TLSConfig  has been added to the cache.

**Comment by dongjiang1989** [comment]
> Could you also add nil checks in `addSafeTLStoYaml` so that we can skip checking for `nil` TLS outside of it? You can also do it in a follow-up PR if you like. WDYT?

Thanks. Got it.  Maybe i will do it  in #6736

**Comment by mviswanathsai** [comment]
Ah! Sorry for the late reply. That would need to be done by @simonpasquier @slashpai or @ArthurSens.

**Comment by simonpasquier** [comment]
@dongjiang1989 LGTM but you need to resolve conflicts now (sorry about that!).

**Comment by dongjiang1989** [comment]
> @dongjiang1989 LGTM but you need to resolve conflicts now (sorry about that!).

Thanks for your review. Fixed it done.

**Review by simonpasquier — APPROVED** [review]
thanks!

**Pull request merged**